packages/next/src/build/webpack/loaders/next-flight-loader/module-proxy.ts

```diff
@@ -28,61 +28,180 @@ const proxyHandlers: ProxyHandler<object> = {
+        // have the Flight runtime extract the inner target instead.
+        return target.$$typeof
+      case 'filepath':
+        return target.filepath
+      case 'name':
+        return target.name
+      case 'async':
+        return target.async
+      // We need to special case this because createElement reads it if we pass this
+      // reference.
+      case 'defaultProps':
+        return undefined
+      case 'getDefaultProps':
+        return undefined
+      // Avoid this attempting to be serialized.
+      case 'toJSON':
+        return undefined
+      case Symbol.toPrimitive.toString():
+        // @ts-ignore
+        return Object.prototype[Symbol.toPrimitive]
       case '__esModule':
         // Something is conditionally checking which export to use. We'll pretend to be
         // an ESM compat module but then we'll check again on the client.
-        target.default = {
-          $$typeof: MODULE_REFERENCE,
-          filepath: target.filepath,
-          // This a placeholder value that tells the client to conditionally use the
-          // whole object or just the default export.
-          name: '',
-          async: target.async,
-        }
+        const moduleId = target.filepath
+        target.default = Object.defineProperties(
+          function () {
+            throw new Error(
+              `Attempted to call the default export of ${moduleId} from the server` +
+                `but it's on the client. It's not possible to invoke a client function from ` +
+                `the server, it can only be rendered as a Component or passed to props of a` +
+                `Client Component.`
+            )
+          },
+          {
+            // This a placeholder value that tells the client to conditionally use the
+            // whole object or just the default export.
+            name: { value: '' },
+            $$typeof: { value: CLIENT_REFERENCE },
+            filepath: { value: target.filepath },
+            async: { value: target.async },
+          }
+        )
         return true
       case 'then':
+        if (target.then) {
+          // Use a cached value
+          return target.then
+        }
         if (!target.async) {
           // If this module is expected to return a Promise (such as an AsyncModule) then
           // we should resolve that with a client reference that unwraps the Promise on
           // the client.
-          const then = function then(
-            resolve: (res: any) => void,
-            _reject: (err: any) => void
-          ) {
-            const moduleReference: Record<string, any> = {
-              $$typeof: MODULE_REFERENCE,
-              filepath: target.filepath,
-              name: '*', // Represents the whole object instead of a particular import.
-              async: true,
+
+          const innerModuleId = target.filepath
+          const clientReference: Function = Object.defineProperties(
+            function () {
```

Comment:
This probably have to be turned back into an object as well for the same reason.